python/ray/tune/ray_trial_executor.py

```diff
@@ -252,27 +280,69 @@ def _setup_remote_runner(self, trial):
             return existing_runner
 
-        if self._cached_actor:
+        if self._cached_actor_pg[0]:
             logger.debug("Cannot reuse cached runner {} for new trial".format(
-                self._cached_actor))
-            with self._change_working_directory(trial):
-                pg = self._pg_manager.clean_trial_placement_group(trial)
+                self._cached_actor_pg[0]))
+            existing_runner, pg = self._cached_actor_pg
+
+            if not self._pg_manager.return_cached_pg(pg, replace_pending=True):
+                # If we could not replace a pending pg, it might not be needed
+                # anymore. Remove instead.
+                self._pg_manager.clean_cached_pg(pg)
+            else:
+                pg = None
 
+            with self._change_working_directory(trial):
                 self._trial_cleanup.add(
-                    trial, actor=self._cached_actor, placement_group=pg)
-            self._cached_actor = None
+                    trial, actor=existing_runner, placement_group=pg)
+            self._cached_actor_pg = (None, None)
+
+        trainable_cls = trial.get_trainable_cls()
+        if not trainable_cls:
+            raise AbortTrialExecution(
+                f"Invalid trainable: {trial.trainable_name}. If you passed "
+                f"a string, make sure the trainable was registered before.")
+        _actor_cls = _class_cache.get(trainable_cls)
 
-        _actor_cls = _class_cache.get(trial.get_trainable_cls())
         if trial.uses_placement_groups:
-            if not self._pg_manager.has_ready(trial.placement_group_factory):
+            if not self._pg_manager.has_ready(trial, update=True):
                 if trial not in self._staged_trials:
-                    if self._pg_manager.stage_trial_pg(
-                            trial.placement_group_factory):
+                    if self._pg_manager.stage_trial_pg(trial):
                         self._staged_trials.add(trial)
+                        self._just_staged_trials.add(trial)
+
+                just_staged = trial in self._just_staged_trials
+
+                # This part of the code is mostly here for testing
+                # purposes. If self._wait_for_pg is set, we will wait here
+                # for that many seconds until the placement group is ready.
+                # This ensures that the trial can be started right away and
+                # not just in the next step() of the trial runner.
+                # We only do this if we have reason to believe that resources
+                # will be ready, soon, i.e. when a) we just staged the PG,
+                # b) another trial just exited, freeing resources, or c)
+                # when there are no currently running trials.
+                if self._wait_for_pg is not None and (
+                        just_staged or self._trial_just_finished_before
+                        or not self.get_running_trials()):
```

Comment:
since this is testing only, would it be possible to simplify the code and remove the `just_staged`/`just_finished_before` concepts and always wait_for_pg/break when ready?

Reply:
This will blow up some test times, e.g. if we only have resources for 2 trials but request 10, there's pending actors most of the time and we will always (i.e. at each `step()`) try to schedule it and wait for n seconds.
I'll try to think of a better solution here